File: ASP.NET Web Forms Application Using Entity Framework Database First - V2/[VB]-ASP.NET Web Forms Application Using Entity Framework Database First - V2/VB/ContosoUniversity.Tests/UnitTest1.vb
Imports System.Text
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports ContosoUniversity

Namespace ContosoUniversity.Tests
	''' <summary>
	''' Summary description for UnitTest1
	''' </summary>
	<TestClass>
	Public Class UnitTest1
		Public Sub New()
			'
			' TODO: Add constructor logic here
			'
		End Sub

		Private testContextInstance As TestContext

		''' <summary>
		'''Gets or sets the test context which provides
		'''information about and functionality for the current test run.
		'''</summary>
		Public Property TestContext() As TestContext
			Get
				Return testContextInstance
			End Get
			Set(ByVal value As TestContext)
				testContextInstance = value
			End Set
		End Property

		#Region "Additional test attributes"
		'
		' You can use the following additional attributes as you write your tests:
		'
		' Use ClassInitialize to run code before running the first test in the class
		' [ClassInitialize()]
		' public static void MyClassInitialize(TestContext testContext) { }
		'
		' Use ClassCleanup to run code after all tests in a class have run
		' [ClassCleanup()]
		' public static void MyClassCleanup() { }
		'
		' Use TestInitialize to run code before running each test 
		' [TestInitialize()]
		' public void MyTestInitialize() { }
		'
		' Use TestCleanup to run code after each test has run
		' [TestCleanup()]
		' public void MyTestCleanup() { }
		'
		#End Region

		Private Function CreateSchoolBL() As SchoolBL
			Dim schoolRepository = New MockSchoolRepository()
			Dim schoolBL = New SchoolBL(schoolRepository)
			schoolBL.InsertDepartment(New Department() With {.Name = "First Department", .DepartmentID = 0, .Administrator = 1, .Person = New Instructor () With {.FirstMidName = "Admin", .LastName = "One"}})
			schoolBL.InsertDepartment(New Department() With {.Name = "Second Department", .DepartmentID = 1, .Administrator = 2, .Person = New Instructor() With {.FirstMidName = "Admin", .LastName = "Two"}})
			schoolBL.InsertDepartment(New Department() With {.Name = "Third Department", .DepartmentID = 2, .Administrator = 3, .Person = New Instructor() With {.FirstMidName = "Admin", .LastName = "Three"}})
			Return schoolBL
		End Function

		<TestMethod, ExpectedException(GetType(DuplicateAdministratorException))>
		Public Sub AdministratorAssignmentRestrictionOnInsert()
			Dim schoolBL = CreateSchoolBL()
			schoolBL.InsertDepartment(New Department() With {.Name = "Fourth Department", .DepartmentID = 3, .Administrator = 2, .Person = New Instructor() With {.FirstMidName = "Admin", .LastName = "Two"}})
		End Sub

		<TestMethod, ExpectedException(GetType(DuplicateAdministratorException))>
		Public Sub AdministratorAssignmentRestrictionOnUpdate()
			Dim schoolBL = CreateSchoolBL()
			Dim origDepartment = (
			    From d In schoolBL.GetDepartments()
			    Where d.Name = "Second Department"
			    Select d).First()
			Dim department = (
			    From d In schoolBL.GetDepartments()
			    Where d.Name = "Second Department"
			    Select d).First()
			department.Administrator = 1
			schoolBL.UpdateDepartment(department, origDepartment)
		End Sub
	End Class
End Namespace

File: ASP.NET Web Forms Application Using Entity Framework Database First - V2/[C#]-ASP.NET Web Forms Application Using Entity Framework Database First - V2/C#/ContosoUniversity.Tests/UnitTest1.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using ContosoUniversity.BLL;
using ContosoUniversity.DAL;

namespace ContosoUniversity.Tests
{
    /// <summary>
    /// Summary description for UnitTest1
    /// </summary>
    [TestClass]
    public class UnitTest1
    {
        public UnitTest1()
        {
            //
            // TODO: Add constructor logic here
            //
        }

        private TestContext testContextInstance;

        /// <summary>
        ///Gets or sets the test context which provides
        ///information about and functionality for the current test run.
        ///</summary>
        public TestContext TestContext
        {
            get
            {
                return testContextInstance;
            }
            set
            {
                testContextInstance = value;
            }
        }

        #region Additional test attributes
        //
        // You can use the following additional attributes as you write your tests:
        //
        // Use ClassInitialize to run code before running the first test in the class
        // [ClassInitialize()]
        // public static void MyClassInitialize(TestContext testContext) { }
        //
        // Use ClassCleanup to run code after all tests in a class have run
        // [ClassCleanup()]
        // public static void MyClassCleanup() { }
        //
        // Use TestInitialize to run code before running each test 
        // [TestInitialize()]
        // public void MyTestInitialize() { }
        //
        // Use TestCleanup to run code after each test has run
        // [TestCleanup()]
        // public void MyTestCleanup() { }
        //
        #endregion

        private SchoolBL CreateSchoolBL()
        {
            var schoolRepository = new MockSchoolRepository();
            var schoolBL = new SchoolBL(schoolRepository);
            schoolBL.InsertDepartment(new Department() { Name = "First Department", DepartmentID = 0, Administrator = 1, Person = new Instructor () { FirstMidName = "Admin", LastName = "One" } });
            schoolBL.InsertDepartment(new Department() { Name = "Second Department", DepartmentID = 1, Administrator = 2, Person = new Instructor() { FirstMidName = "Admin", LastName = "Two" } });
            schoolBL.InsertDepartment(new Department() { Name = "Third Department", DepartmentID = 2, Administrator = 3, Person = new Instructor() { FirstMidName = "Admin", LastName = "Three" } });
            return schoolBL;
        }

        [TestMethod]
        [ExpectedException(typeof(DuplicateAdministratorException))]
        public void AdministratorAssignmentRestrictionOnInsert()
        {
            var schoolBL = CreateSchoolBL();
            schoolBL.InsertDepartment(new Department() { Name = "Fourth Department", DepartmentID = 3, Administrator = 2, Person = new Instructor() { FirstMidName = "Admin", LastName = "Two" } });
        }

        [TestMethod]
        [ExpectedException(typeof(DuplicateAdministratorException))]
        public void AdministratorAssignmentRestrictionOnUpdate()
        {
            var schoolBL = CreateSchoolBL();
            var origDepartment = (from d in schoolBL.GetDepartments()
                                  where d.Name == "Second Department"
                                  select d).First();
            var department = (from d in schoolBL.GetDepartments()
                              where d.Name == "Second Department"
                              select d).First();
            department.Administrator = 1;
            schoolBL.UpdateDepartment(department, origDepartment);
        }
    }
}
